Add resource types to allowed types for solution-template-validation-tests

stuartleeks opened the pull request:
### Best Practice Checklist
Check these items before submitting a PR... See the Contribution Guide for the full detail: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md 

1. uri's compatible with all clouds (Stack, China, Government)
1. Staged artifacts use _artifactsLocation & _artifactsLocationSasToken
1. Use resourceGroup().location for resource locations
1. Folder names for artifacts (nestedtemplates, scripts, DSC)
1. Use literal values for apiVersion (no variables)
1. Parameter files (GEN-UNIQUE for value generation and no "changemeplease" values
1. $schema and other uris use https
1. Use uniqueString() whenever possible to generate names for resources.  While this is not required, it's one of the most common failure points in a deployment. 
1. Update the metadata.json with the current date

For details: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/bp-checklist.md

- [X] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

### Changelog

* Add Microsoft.Insights/autoscaleSettings and Microsoft.Authorization/roleAssignments to the allowed resource types in solution-template-validation-tests

stuartleeks commented:
@bmoore-msft these are the changes we discussed yesterday :-)

bmoore-msft commented on test/solution-template-validation-tests/test/mainTemplateTests.js:
```diff
@@ -138,10 +138,13 @@ describe('mainTemplate.json file - ', () => {
                     var val = templateObject.resources[resource];
                     if (val.type) {
                         var rType = val.type.split('/');
+                        var lowerType = val.type.toLowerCase();
```
Can we make this consistent across the checks?  i.e. use rType[0].toLowerCase() or lowerType but not both?

stuartleeks replied:
There are two class of tests here: anything from an RP (`Microsoft.Compute/*`), and specific resource type (`Microsoft.Insights/autoscaleSettings`).

I could change the new tests to check `rType[0]` and `rType[1]`? Or convert the existing `rType[0]` check into `lowerType.startsWith('microsoft.compute/')`?

Thoughts?

stuartleeks replied:
Or, how about having a data structure with a specification of the allowed types and have the condition look up in that? Might make it clearer going forwards as more types are added?

E.g. 
```javascript
            var allowedResourceTypes = {
                'microsoft.compute' : ['*'],
                'microsoft.storage' : ['*'],
                'microsoft.network' : ['*'],
                'microsoft.resources' : ['*'],
                'microsoft.insights' : ['autoscalesettings'],
                'microsoft.authorization' : ['roleassignments']
            };
```

stuartleeks commented:
@bmoore-msft  I've just pushed a commit that reworks the test as per my last comment. Let me know your thoughts. 

Also happy to squash into a single commit - lmk.

bmoore-msft commented:
I like this - very readable (easy to augment)...

@vivsriaus can you have a look?

vivsriaus commented on test/solution-template-validation-tests/test/mainTemplateTests.js:
```diff
@@ -130,19 +130,31 @@ describe('mainTemplate.json file - ', () => {
-                        expect(cond, getErrorMessage(val, templateJSONObject.filepath, message + '. The resource object: ' + JSON.stringify(val))).to.be.true;
+                        var rType = val.type.toLowerCase().split('/');
+                        var providerLower = rType[0];
+                        var typeLower = rType[1];
```
null check for rType[1] please

stuartleeks replied:
Done

vivsriaus replied:
@stuartleeks I was referring to the below line for NRE. I know there are slim chances the resourceType contains just the provider namespace (and not include the resourceType as well), but rType[1] will throw when "/" is absent in type.
`
var typeLower = rType[1];
`

vivsriaus commented on test/solution-template-validation-tests/test/mainTemplateTests.js:
```diff
@@ -130,19 +130,31 @@ describe('mainTemplate.json file - ', () => {
             templateFileJSONObjects.forEach(templateJSONObject => {
                 var templateObject = templateJSONObject.value;
                 templateObject.should.have.property('resources');
                 var message = 'in file:' + templateJSONObject.filename + ' is NOT a compute, network or a storage resource type';
```
Should now include Insights and Authorization as well

vivsriaus commented on test/solution-template-validation-tests/test/mainTemplateTests.js:
```diff
@@ -130,19 +130,31 @@ describe('mainTemplate.json file - ', () => {
         });
 
         it('template must contain only approved resources (Compute/Network/Storage)', () => {
```
Please update test description to include Insights and AuthN

stuartleeks commented:
@vivsriaus I've added the guard on the array access. For the output message, I updated that to ensure that it stays in sync as the allowed types list is edited (and in the same spirit made the test name more generic)

Let me know if you're happy with these changes :-)
